File: 2709-greatest-common-divisor-traversal/2709-greatest-common-divisor-traversal.java
class Solution {
    public int n;
    public int[] p;
    public int[] count;
    
    public int find(int x) {
        if (p[x] == x) {
            return x;
        }
        
        return p[x] = find(p[x]);
    }
    
    public void union(int a, int b) {
        if (a == b) {
            return;
        }
        
        if (count[a] < count[b]) {
            int temp = a;
            a = b;
            b = temp;
        }
        
        p[b] = a;
        count[a] += count[b];
    }
    
    public boolean canTraverseAllPairs(int[] nums) {
        n = nums.length;
        p = new int[n];
        count = new int[n];
        
        for (int i = 0; i < n; i++) {
            p[i] = i;
            count[i] = 1;
        }
        
        Map<Integer, Integer> m = new HashMap<>();
        for (int i = 0; i < n; i++) {
            for (int div = 2; div * div <= nums[i]; div++) {
                if (nums[i] % div == 0) {
                    if (m.containsKey(div)) {
                        int a = find(i);
                        int b = find(m.get(div));
                        
                        if (a != b) {
                            union(a, b);
                        }
                    } else {
                        m.put(div, i);
                    }
                    
                    while (nums[i] % div == 0) {
                        nums[i] /= div;
                    }
                }
            }
            
            if (nums[i] != 1) {
                if (m.containsKey(nums[i])) {
                    int a = find(i);
                    int b = find(m.get(nums[i]));
                    
                    if (a != b) {
                        union(a, b);
                    }
                } else {
                    m.put(nums[i], i);
                }
            }
        }
        
        return count[find(0)] == n;
    }
}

File: 2709-greatest-common-divisor-traversal/2709-greatest-common-divisor-traversal.cpp
class Solution {
public:
  typedef vector<int> vi;
  typedef vector<vi> vii;
  int n;
  vi count;
  vi parent;
  unordered_map<int,int>m;
  int find(int x){
    if(parent[x]==x){
      return x;
    }
    return parent[x]=find(parent[x]);
  }
  void uni(int a,int b){
    parent[b]=a;
    count[a]+=count[b];
  }
  bool canTraverseAllPairs(vector<int>& nums) {
    n=nums.size();
    parent=vi(n,0);
    count=vi(n,1);
    for(int i=0;i<n;i++){
      parent[i]=i;
    }
    for(int i=0;i<n;i++){
      for(int j=2;j*j<=nums[i];j++){
        if(nums[i]%j==0){
          if(m.find(j)!=m.end()){
            int a=find(i);
            int b=find(m[j]);
            if(a!=b){
              uni(a,b);
            }
          }else{
            m[j]=i;
          }
          while(nums[i]%j==0){
            nums[i]/=j;
          }
        }
      }
      if(nums[i]!=1){
        if(m.find(nums[i])!=m.end()){
          int a=find(i);
          int b=find(m[nums[i]]);
          if(a!=b){
            uni(a,b);
          }
        }else{
          m[nums[i]]=i;
        }
      }
    }
    return count[find(0)]==n;
  }
};